subsys/fs/zms/zms.c

```diff
@@ -0,0 +1,1804 @@
+	rc = zms_flash_ate_rd(fs, close_addr + ate_size, empty_ate);
+	if (rc) {
+		return rc;
+	}
+
+	return 0;
+}
+
+/**
+ * @brief Helper to find an ATE using its ID
+ *
+ * @param fs Pointer to file system
+ * @param id Id of the entry to be found
+ * @param start_addr Address from where the search will start
+ * @param end_addr Address where the search will stop
+ * @param ate pointer to the found ATE if it exists
+ * @param ate_addr Pointer to the address of the found ATE
+ *
+ * @retval 0 No ATE is found
+ * @retval 1 valid ATE with same ID found
+ * @retval < 0 An error happened
+ */
+static int zms_find_ate_with_id(struct zms_fs *fs, uint32_t id, uint64_t start_addr,
+				uint64_t end_addr, struct zms_ate *ate, uint64_t *ate_addr)
+{
+	int rc;
+	int previous_sector_num = ZMS_INVALID_SECTOR_NUM;
+	uint64_t wlk_prev_addr, wlk_addr;
+	int prev_found = 0;
+	struct zms_ate wlk_ate;
+	uint8_t current_cycle;
+
+	wlk_addr = start_addr;
+
+	do {
+		wlk_prev_addr = wlk_addr;
+		rc = zms_prev_ate(fs, &wlk_addr, &wlk_ate);
+		if (rc) {
+			return rc;
+		}
+		if (wlk_ate.id == id) {
+			/* read the ate cycle only when we change the sector or if it is
+			 * the first read ( previous_sector_num == ZMS_INVALID_SECTOR_NUM).
+			 */
+			rc = zms_get_cycle_on_sector_change(fs, wlk_prev_addr, previous_sector_num,
+							    &current_cycle);
+			if (rc) {
+				return rc;
+			}
+			if (zms_ate_valid_different_sector(fs, &wlk_ate, current_cycle)) {
+				prev_found = 1;
+				break;
+			}
+			previous_sector_num = SECTOR_NUM(wlk_prev_addr);
+		}
+	} while (wlk_addr != end_addr);
+
+	*ate = wlk_ate;
+	*ate_addr = wlk_prev_addr;
+
+	return prev_found;
+}
+
+/* garbage collection: the address ate_wra has been updated to the new sector
+ * that has just been started. The data to gc is in the sector after this new
+ * sector.
+ */
+static int zms_gc(struct zms_fs *fs)
+{
+	int rc, sec_closed;
+	struct zms_ate close_ate, gc_ate, wlk_ate, empty_ate;
+	uint64_t sec_addr, gc_addr, gc_prev_addr, wlk_addr, wlk_prev_addr, data_addr, stop_addr;
+	uint8_t previous_cycle = 0;
+	size_t ate_size;
+
+	ate_size = zms_al_size(fs, sizeof(struct zms_ate));
+
+	rc = zms_get_sector_cycle(fs, fs->ate_wra, &fs->sector_cycle);
+	if (rc == -ENOENT) {
+		/* Erase this new unused sector if needed */
+		rc = zms_flash_erase_sector(fs, fs->ate_wra);
+		if (rc) {
+			return rc;
+		}
+		/* sector never used */
+		rc = zms_add_empty_ate(fs, fs->ate_wra);
+		if (rc) {
+			return rc;
+		}
+		/* At this step we are sure that empty ATE exist.
+		 * If not, then there is an I/O problem.
+		 */
+		rc = zms_get_sector_cycle(fs, fs->ate_wra, &fs->sector_cycle);
+		if (rc) {
+			return rc;
+		}
+	} else if (rc) {
+		/* bad flash read */
+		return rc;
+	}
+	previous_cycle = fs->sector_cycle;
+
+	sec_addr = (fs->ate_wra & ADDR_SECT_MASK);
+	zms_sector_advance(fs, &sec_addr);
+	gc_addr = sec_addr + fs->sector_size - ate_size;
+	/* skip header ATEs */
+	stop_addr = gc_addr - 2 * ate_size;
```

Comment:
Minor: This line could be moved to just before do{} as it is not needed before that.